### ISSUE TITLE
Request: Multi-file diff

### ISSUE THREAD

**Issue by sbwoodside** [ForkIssues/Tracker#261]
I'm often reviewing similar changes across a bunch of files, and it would be easier to maintain context if I could view multiple files at once. I envision that they would be stacked in the file viewing panel. I make a really quick mockup which is attached.

This would allow me to stage similar chunks from multiple files more quickly, and to view changes that have a cross-file impact more quickly.

I envision that you be able to select multiple files or select a folder in tree view to view all of the files in that folder.

<img width="726" alt="multi-file mockup" src="https://user-images.githubusercontent.com/47834/38584889-d51be798-3ce5-11e8-8cd4-f10489797ac8.png">

**Comment by filiphr** [comment]
I think that this is one of the only features that I am missing from git-fork compared to SourceTree. Although git-fork is really fast when switching between files it would still be nice to have such option 😄

**Comment by DanPristupov** [comment]
It's very unlikely this feature will appear in the near future. There are multiple problems which will require major changes:

* performance
* memory usage
* minimap on scrollbar must be reworked
* increasing of general code complexity

**Comment by filiphr** [comment]
I completely understand.

Thanks for taking the time for reply

**Comment by relgit** [comment]
This feature is frankly the only reason I'm still using Sourcetree. If you add it, Fork will become an insta-buy for me.

**Comment by IlyaSemenov** [comment]
This feature is a must when dealing with mass refactoring. There could be literally hundreds of slightly changed files (e.g. renamed import or simply moved without any changes at all), yet some of files could accidentally have residual unwanted changes so everything must be screened before making a commit.

Currently, the process is:
1. Select the first file in Unstaged
2. Ensure that the changes are very minimal as expected, Press Enter
3. The selection will change to the next file automatically
4. Go to step 2 and repeat all over

Not only it's tiresome to go through files one by one, the problem is that the brain starts to "predict" that it must press Enter, and more often than not presses it too early when there's suddenly a file with unexpected change. But the file is gone to "Staged" section already, and it's hard to find it again to actually review it.

If one could select the entire directory in "Unstaged" and see the complete diff under that directory, that would a huge time saver.

If that's about performance, that could be worked around with explicit "View diff" button that must be pressed:
<img width="1020" alt="Image" src="https://github.com/user-attachments/assets/53e344f7-031f-481e-ba3a-14f813ef9758" />
Here, under the large Fork icon in the right section which is currently useless, there could be a new button like "View diff", perhaps with some disclaimer that it's time consuming as it will have to prepare a diff of 1234 files.

**Comment by Memphis264** [comment]
Hello, just as for @relgit and others, this is a must feature for me and the one keeping me from using overall great Fork. @DanPristupov do the issues blocking implementing this feature still apply today?

Being able to select some/all staged/unstaged files and see their diff seamlessly is really great for checking related changes across files. Also it is practical to be able to select all files and just scroll through them before staging them.

I'm sharing SourceTree example and also Fork commit overview feature which shows diff of all files (after clicking "Expand All" button) - it's similar feature just for commited changes rather than local changes..

Would it be possible to reconsider implementing this please?

Thanks!

<img width="1309" height="994" alt="Image" src="https://github.com/user-attachments/assets/9c9b9bad-b61c-4afd-a000-cff855c33920" />
<img width="1547" height="1266" alt="Image" src="https://github.com/user-attachments/assets/1e6ab5a6-837f-4997-bc8f-632d141def6b" />

**Comment by DanPristupov** [comment]
It's not as easy as it sounds. There are a few tricky spots here.

1. Technical. If I remember correctly, Windows doesn't have a proper virtual ListView with variable element height support which is required for this kind of layout.
2. The design is hard to integrate into the current layout.
- The mechanism for how these files will be selected is unclear. Simple selection is pointless. You'd have to keep hitting Ctrl+A. That's not right. That would also be error prone (if a newly modified file appears in an off-screen area, it won't be selected). Right after the feature release, people would not want to press Ctrl+A and would ask to have "select all" by default. This is an uncommon behavior for the traditional [list-view -> details] layout.
TL;DR: I don't know how to make such a list a complete feature: fast, cross-platform, intuitive, and self-explanatory.

**Comment by niorko** [comment]
@DanPristupov Both points make sense, but I keep wondering about a different framing:

1. Does the multi-file view actually need to be a ListView of file items? If it were one flat virtual document where the item is a *line* (diff line / hunk separator / file header just being different line kinds), everything would be fixed-height and the variable-height problem wouldn't exist on either platform. I'm guessing the current diff view already virtualizes lines to handle huge files - if so, wouldn't a multi-file diff be more or less the same surface with header lines interleaved? I suspect that's how Sublime Merge pulls it off: it feels like a text surface rather than a list of file widgets.

2. On selection - you're right that selection-based multi-view is broken (the off-screen new file case especially). But what if it weren't selection-based at all? E.g. clicking a folder node (or the "Unstaged" header) shows the combined diff of everything under it, as a live view rather than a frozen selection. A new file would just appear. And it might even fit the list-view -> details layout naturally: a file's detail is its diff, so a folder's detail could be the combined diff of its contents. No Ctrl+A, no new selection semantics.

No idea how well this maps onto Fork's internals, of course - just thinking out loud...

**Comment by DanPristupov** [comment]
> Does the multi-file view actually need to be a ListView of file items? 

Are you suggesting to replace text control with a ListView of lines?
Then we would need to implement text selection,  keyboard and mouse handling, word wrapping, performance for minified files (4Mb as a single line), and other basic native text field features.

> If it were one flat virtual document where the item is a line (diff line / hunk separator / file header just being different line kinds), everything would be fixed-height and the variable-height problem wouldn't exist on either platform.

Fork has not only a text diff but also a binary diff, an image diff, a dynamic LFS diff, and even a submodule diff (which has sub-scrolling).


> But what if it weren't selection-based at all? E.g. clicking a folder node (or the "Unstaged" header) shows the combined diff of everything under it

Good, this solves the 'new file' problem and is actually intuitive. But how would the user see all changes at once? There is no empty "root" item.
As a related problem, this will not work for the flat-file list layout.

**Comment by filiphr** [comment]
>1. Technical. If I remember correctly, Windows doesn't have a proper virtual ListView with variable element height support which is required for this kind of layout.

@DanPristupov you are mentioning Windows. How about Mac? If Windows does not have a way to support this properly and Mac has, is it an option to provide the feature for Mac only?


>As a related problem, this will not work for the flat-file list layout.

In the flat-file list layout on Mac you can do CMD + Click to select the files you want

<img width="814" height="782" alt="Image" src="https://github.com/user-attachments/assets/c31dbeb6-0e2f-4c84-8588-2c01d0a2e119" />

At the moment only the first selected file is show in the diff view on the right hand side. Isn't it possible to use a similar component like in the Commit tab also in this right hand side view like @Memphis264 mentioned?

and as @niorko said, I also have no idea how well this maps onto Fork's internals, just thinking out loud...
